pyiceberg/table/__init__.py

```diff
@@ -1904,3 +2001,144 @@ def _generate_snapshot_id() -> int:
+
+    # This is an iter, so we don't have to materialize everything every time
+    # This will be more relevant when we start doing partitioned writes
+    yield from write_file(table, iter([WriteTask(df)]))
+
+
+def _manifests_to_manifest_list(
+    table: Table, snapshot_id: int, parent_snapshot_id: Optional[int], manifests: List[ManifestFile], summary: Dict[str, str]
+) -> Snapshot:
+    manifest_list_file_path = f'{table.location()}/metadata/{_generate_manifest_list_filename(snapshot_id=snapshot_id)}'
+    with write_manifest_list(
+        format_version=table.metadata.format_version,
+        output_file=table.io.new_output(manifest_list_file_path),
+        snapshot_id=snapshot_id,
+        parent_snapshot_id=parent_snapshot_id,
+        sequence_number=None,
+    ) as writer:
+        writer.add_manifests(manifests)
+
+    return Snapshot(
+        snapshot_id=snapshot_id,
+        parent_snapshot_id=parent_snapshot_id,
+        manifest_list=manifest_list_file_path,
+        sequence_number=table._next_sequence_number(),
+        summary=summary,
+        schema_id=table.schema().schema_id,
+    )
+
+
+class _AppendManifest(ABC):
+    @abstractmethod
+    def append_manifest(self, manifest_file: ManifestFile) -> _AppendManifest:
+        pass
+
+    @abstractmethod
+    def manifests(self) -> List[ManifestFile]:
+        pass
+
+
+class _MergeAppend:
+    _table: Table
+    _snapshot_id: int
+    _added_manifests: List[ManifestFile]
+    _added_datafiles: List[DataFile]
+
+    def __init__(self, table: Table, snapshot_id: int) -> None:
+        self._table = table
+        self._snapshot_id = snapshot_id
+        self._added_manifests = []
+        self._added_datafiles = []
+
+    def append_datafile(self, data_file: DataFile) -> _MergeAppend:
+        self._added_datafiles.append(data_file)
+        return self
+
+    def append_manifest(self, manifest_file: ManifestFile) -> _MergeAppend:
+        if manifest_file.content != ManifestContent.DATA:
+            raise ValueError(f"Can only add DATA manifests: {manifest_file.content}")
+
+        if manifest_file.added_snapshot_id is not None and self._snapshot_id != manifest_file.added_snapshot_id:
+            # We have to rewrite the manifest because there is a new snapshot
+            self._added_manifests.append(self._copy_manifest(manifest_file))
+        else:
+            self._added_manifests.append(manifest_file)
+
+        return self
+
+    def _copy_manifest(self, manifest_file: ManifestFile) -> ManifestFile:
+        """Rewrites a manifest file with a new snapshot-id.
+
+        Args:
+            manifest_file: The existing manifest file
+
+        Returns:
+            New manifest file with the current snapshot-id
+        """
+        with write_manifest(
+            format_version=self._table.format_version,
+            spec=self._table.specs()[manifest_file.partition_spec_id],
+            # TODO: We want to get the schema from the metadata
```

Comment:
I think this should be fine. I don't think that we copy the schema from the old manifest in Java either.

Reply:
That's good, thanks! 👍